core/src/main/java/com/kin/ecosystem/core/accountmanager/AccountManagerImpl.java

```diff
@@ -174,6 +181,55 @@ public void onFailure(KinEcosystemException exception) {
 	public void switchAccount(final int accountIndex, @NonNull final KinCallback<Boolean> callback) {
 		Logger.log(new Log().withTag(TAG).put("switchAccount", "start"));
 		final String address = blockchainSource.getPublicAddress(accountIndex);
+		updateWalletAddressProcess(address, accountIndex, callback);
+	}
+
+	private void updateWalletAddressProcess(final String address, final int accountIndex,
+		final KinCallback<Boolean> callback) {
+		blockchainSource.getMigrationInfo(address, new Callback<MigrationInfo, ApiException>() {
+			@Override
+			public void onResponse(MigrationInfo migrationInfo) {
+				boolean isRestorable = migrationInfo.isRestorable();
+				Logger.log(new Log().withTag(TAG).put("switchAccount", "isRestorable = " + isRestorable));
+				if (isRestorable) {
+					startMigrationProcess(migrationInfo, address, accountIndex, callback);
+				} else {
+					// TODO: 02/04/2019 is there another normal to send that error to onFailure?
+					onFailure(new ApiException(ServiceException.WALLET_WAS_NOT_CREATED_IN_THIS_APP,
```

Comment:
I think you can move the onFailure login to a method that accepts `KinEcosystemException` and the callback.
So you won't need to create the ApiException